dali/pipeline/operator/sequence_shape.h

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATOR_SEQUENCE_SHAPE_H_
+#define DALI_PIPELINE_OPERATOR_SEQUENCE_SHAPE_H_
+
+#include <string>
+#include <type_traits>
+#include <vector>
+#include "dali/pipeline/data/tensor.h"
+#include "dali/pipeline/data/views.h"
+
+namespace dali {
+
+struct LayoutDesc {
+  static inline LayoutDesc FrameAndChannel(const TensorLayout &layout) {
+    int frame_dim = VideoLayoutInfo::FrameDimIndex(layout);
+    int channel_dim = ImageLayoutInfo::ChannelDimIndex(layout);
+    frame_dim = frame_dim > 1 ? -1 : frame_dim;
+    channel_dim = channel_dim > 1 ? -1 : channel_dim;
+    if (frame_dim != 0 && channel_dim != 0) {
+      return {};
+    }
+    return {layout, frame_dim, channel_dim};
+  }
+
+  static inline LayoutDesc Frame(const TensorLayout &layout) {
+    if (VideoLayoutInfo::FrameDimIndex(layout) != 0) {
+      return {};
+    }
+    return {layout, 0};
+  }
+
+  inline int NumDims() const {
+    return (frame_dim >= 0) + (channel_dim >= 0);
+  }
+
+  TensorLayout layout = {};
+  int frame_dim = -1;
+  int channel_dim = -1;
+};
+
+class ExpandDesc {
+ public:
+  inline ExpandDesc(const TensorListShape<> &shape, const LayoutDesc &layout_desc)
+      : layout_desc_{layout_desc},
+        has_channels_{layout_desc_.channel_dim >= 0},
+        has_frames_{layout_desc_.frame_dim >= 0},
+        is_channel_first_{has_channels_ && layout_desc_.channel_dim < layout_desc_.frame_dim},
+        num_dims_{has_channels_ + has_frames_},
+        expanded_shape_{shape.first(num_dims_)},
+        num_elements_{expanded_shape_.num_elements()} {}
```

Comment:
What's missing here is an explicit indication whether channels should or should not be expanded.

Reply:
I'll rename `has_channels` to `expand_channels`, if we don't want to expand channels they are just ignored and has channels is false. There is no need to know if the layout actually has channels if we do not plan on expanding them.

Reply:
Done: there are now expand_channels and expand_frames fields.